### PR TITLE
expand discussion with examples, details, and references

### LINKED ISSUE #380: JLBP-2: Add Design for inheritance or prohibit it.
Taken straight out of Effective Java Item 17, this is part of minimizing API surface. 

Possibly figure out how to cite things like this in footnotes or some such.

### LINKED ISSUE #381: JLBP-2 expand on Try not to expose types from your dependencies.
We should be a little more explicit about what "exposing" means.

In particular add

* Avoid subclassing types from other libraries
* Do not use types from other libraries as public method arguments or return values

### PR DESCRIPTION
**Pull request by elharo**
fix #380
fix #381 
@garrettjonesgoogle

### CODE REVIEW

File: library-best-practices/JLBP-2.md [hunk 1]
```diff
@@ -2,19 +2,40 @@
...
+
+  - Method arguments and return values from public methods should be standard Java 
+    types such as `java.time.LocalDate` or classes defined in the library itself,
+    not third party types such as `org.joda.time.LocalDate`.
+
+  - Your own public types should not be subclasses or implementations of types
+    in third party libraries.  
+
+  - Third party libraries can change their own API more frequently than you like
+    or do so at inconvenient times. If a third party type on the surface of
```
Comment (garrettjonesgoogle) [inline library-best-practices/JLBP-2.md]: The wording gets a little awkward here ("or do so at...")

File: library-best-practices/JLBP-2.md [hunk 2]
```diff
@@ -2,19 +2,40 @@
...
-    big problem for Guava, which made a major version bump
-    every 6 months, although Guava no longer does this.
-- Use package-protected classes and methods when the API is internal
-  only and should not be used by consumers.
-- Do not mark methods and classes public by default.
-  Assume non-public until a need is known.
+
+  - Method arguments and return values from public methods should be standard Java 
+    types such as `java.time.LocalDate` or classes defined in the library itself,
+    not third party types such as `org.joda.time.LocalDate`.
```
Comment (suztomo) [inline library-best-practices/JLBP-2.md]: This is nice enhancement, first explaining the meaning of "exposing types from your dependencies"

File: library-best-practices/JLBP-2.md [hunk 3]
```diff
@@ -2,19 +2,40 @@
...
+
+  - Your own public types should not be subclasses or implementations of types
+    in third party libraries.  
+
+  - Third party libraries can change their own API more frequently than you like
+    or do so at inconvenient times. If a third party type on the surface of
+    your API changes or is removed, you either have to break your own API or remain with an older,
+    unsupported version of the library that is likely to cause diamond dependency
+    problems. Historically this was a big problem for libraries that exposed Guava types
```
Comment (suztomo) [inline library-best-practices/JLBP-2.md]: Nice. (I was wondering why this Guava is an example of this section)

File: library-best-practices/JLBP-2.md [hunk 4]
```diff
@@ -2,19 +2,40 @@
...
+    such as `com.google.common.io.OutputSupplier` because Guava bumped major versions
```
Comment (suztomo) [inline library-best-practices/JLBP-2.md]: bump -> incremented (to use specific verb)

File: library-best-practices/JLBP-2.md [hunk 5]
```diff
@@ -2,19 +2,40 @@
...
+
 - Prefer fewer packages over more packages to avoid
   unnecessarily publicizing internal details,
-  since any dependency across package boundaries need to be
+  since any dependency across package boundaries needs to be
   public. (We may revisit this when we can rely on
   the new module system in Java 11 or later.) 
+
+
+<b id="item15">1</b> Bloch, Joshua. "Item 15: Minimize the accessibility of classes and members." Effective Java, 3rd Edition. Boston: Addison-Wesley, 2018. p. 73[↩](#a1)
```
Comment (suztomo) [inline library-best-practices/JLBP-2.md]: Would you add line break, as you already did in line 39

File: library-best-practices/JLBP-2.md [hunk 6]
```diff
@@ -2,19 +2,40 @@
...
+
+<b id="item19">2</b> Bloch, Joshua. "Item 19: Design and document for inheritance or else 
+prohibit it." Effective Java, 3rd Edition. Boston: Addison-Wesley, 2018. p. 93[↩](#a2)
```
Comment (suztomo) [inline library-best-practices/JLBP-2.md]:
```suggestion
prohibit it." Effective Java, 3rd Edition. Boston: Addison-Wesley, 2017. p. 93[↩](#a2)
```

> Publication Date: December 18, 2017
https://www.amazon.com/Effective-Java-Joshua-Bloch-ebook/dp/B078H61SCH

File: library-best-practices/JLBP-2.md (hunk 1 above)
Comment (elharo) [reply library-best-practices/JLBP-2.md]: removed. It's redundant.

File: library-best-practices/JLBP-2.md (hunk 4 above)
Comment (elharo) [reply library-best-practices/JLBP-2.md]: done

File: library-best-practices/JLBP-2.md (hunk 5 above)
Comment (elharo) [reply library-best-practices/JLBP-2.md]: done

File: library-best-practices/JLBP-2.md (hunk 6 above)
Comment (elharo) [reply library-best-practices/JLBP-2.md]: The book itself lists 2018. Publishers often bump the copyright year by +1 for books released late in the year. I *think* citations usually go by what's listed in the book.

Comment (suztomo) [reply library-best-practices/JLBP-2.md]: Thank you for clarification.

**Pull request merged**